camel/functions/search_functions.py

```diff
@@ -286,24 +359,61 @@ def search_google_and_summarize(query: str) -> str:
     for item in responses:
         if "url" in item:
             url = item.get("url")
-            # Extract text
-            text = text_extract_from_web(str(url))
+            # Extract text from the web page
+            try:
+                extracted_text = text_extract_from_web(str(url))
+            except Exception:
+                continue
+            # Using chatgpt summarise text
+            answer = summarize_text(extracted_text, query)
+
+            # Let chatgpt decide whether to continue search or not
+            if continue_search(query=query, answer=answer):
+                continue
+            else:
+                return answer
+
+    return "Failed to find the answer from the Google search."
+
+
+def search_duckduckgo_and_summarize(query: str) -> str:
```

Comment:
`search_google_and_summarize`and `search_duckduckgo_and_summarize`share a lot of logic with minor differences in the search source. Abstracting the common logic into one function and passing the search function as an argument could reduce duplication.